Monadic grammar, threaded lexer, `-a` enabled, but no type signature, fails to compile

I noticed that when I have a monadic grammar with a threaded lexer, generated with ~~(just)~~ `-a` enabled, and do not include a type signature on a production, I get type errors in the generated code. I don’t think this is the expected behaviour. Here is an example of such a grammar:

```
-- compileBug.y
%name parse prod

%tokentype { Token }

%monad { P } { bindP } { returnP }
%error { error "parse error" }
%lexer { lexer } { EOF }

%token
  IDENT  { Identifier $$ }

%%

-- Uncommenting the type signature below makes the generated code compile
prod -- :: { () }
  : IDENT { () }

{
data Token = EOF | Identifier String

type P a = String -> (a, String)

bindP :: P a -> (a -> P b) -> P b
bindP p f s = let (x,s') = p s in f x s'

returnP :: a -> P a
returnP = (,)

lexer :: (Token -> P a) -> P a
lexer cont s = cont (case s of { "" -> EOF; _ -> Identifier s }) ""

main = return ()
 }
```

Without the type signature:

    $ happy -a -o compileBug.hs compileBug.y
    $ ghc compileBug.hs
    [1 of 1] Compiling Main             ( compileBug.hs, compileBug.o )
    
    compileBug.hs:61:17: error:
        • Couldn't match type ‘t4’ with ‘()’
          ‘t4’ is a rigid type variable bound by
            the type signature for:
              happyReduce_1 :: forall t4.
                               Int
                               -> Token
                               -> Int
                               -> Happy_IntList
                               -> HappyStk (HappyAbsSyn t4)
                               -> P (HappyAbsSyn t4)
            at compileBug.hs:60:18
          Expected type: Int
                         -> Token
                         -> Int
                         -> Happy_IntList
                         -> HappyStk (HappyAbsSyn t4)
                         -> P (HappyAbsSyn t4)
            Actual type: Int
                         -> Token
                         -> Int
                         -> Happy_IntList
                         -> HappyStk (HappyAbsSyn ())
                         -> P (HappyAbsSyn ())
        • In the expression: happySpecReduce_1 0 happyReduction_1
          In an equation for ‘happyReduce_1’:
              happyReduce_1 = happySpecReduce_1 0 happyReduction_1
        • Relevant bindings include
            happyReduce_1 :: Int
                             -> Token
                             -> Int
                             -> Happy_IntList
                             -> HappyStk (HappyAbsSyn t4)
                             -> P (HappyAbsSyn t4)
              (bound at compileBug.hs:61:1)
    $

With the type signature uncommented:

    $ happy -a -o compileBug.hs compileBug.y
    $ ghc compileBug.hs
    [1 of 1] Compiling Main             ( compileBug.hs, compileBug.o )
    Linking compileBug ...
    $

Note that this is a regression (I have at least some older version of Happy on which the type signature is not required to generate compiling code). This bug seems present in at least ~~1.19.5 and~~ HEAD.

Alright. I did some bisecting (using the above sample file `compileBug.y`) and it looks like the bug was indeed introduced after 1.19.5.

* The last working commit is 6ad73f262d2034028c6fa148cee914263c884ccc.
* Afterwards, there are two commits 00d561cde9706090f46b358f74fcd4908a23be1a and 35a5a19a991afd59ee040daf3577a2c786e7f986 for which the error message is different.

```
$ happy -a -o compileBug.hs compileBug.y
$ ghc compileBug.hs
[1 of 1] Compiling Main             ( compileBug.hs, compileBug.o )

compileBug.hs:60:73: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘HappyStk’, namely ‘HappyAbsSyn’
      In the type signature:
        happyReduce_1 :: Int
                         -> Token
                            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:60:90: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘P’, namely ‘HappyAbsSyn’
      In the type signature:
        happyReduce_1 :: Int
                         -> Token
                            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:89:30: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘P’, namely ‘HappyAbsSyn’
      In the type signature:
        happyParse :: Int -> P HappyAbsSyn

compileBug.hs:91:57: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘HappyStk’, namely ‘HappyAbsSyn’
      In the type signature:
        happyNewToken :: Int
                         -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:91:74: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘P’, namely ‘HappyAbsSyn’
      In the type signature:
        happyNewToken :: Int
                         -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:93:73: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘HappyStk’, namely ‘HappyAbsSyn’
      In the type signature:
        happyDoAction :: Int
                         -> Token
                            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:93:90: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘P’, namely ‘HappyAbsSyn’
      In the type signature:
        happyDoAction :: Int
                         -> Token
                            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn

compileBug.hs:95:102: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘HappyStk’, namely ‘HappyAbsSyn’
      In the second argument of ‘Happy_Data_Array.Array’, namely
        ‘Int
         -> Token
            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn’
      In the type signature:
        happyReduceArr :: Happy_Data_Array.Array Int (Int
                                                      -> Token
                                                         -> Int
                                                            -> Happy_IntList
                                                               -> HappyStk HappyAbsSyn
                                                                  -> P HappyAbsSyn)

compileBug.hs:95:119: error:
    • Expecting one more argument to ‘HappyAbsSyn’
      Expected a type, but ‘HappyAbsSyn’ has kind ‘* -> *’
    • In the first argument of ‘P’, namely ‘HappyAbsSyn’
      In the second argument of ‘Happy_Data_Array.Array’, namely
        ‘Int
         -> Token
            -> Int -> Happy_IntList -> HappyStk HappyAbsSyn -> P HappyAbsSyn’
      In the type signature:
        happyReduceArr :: Happy_Data_Array.Array Int (Int
                                                      -> Token
                                                         -> Int
                                                            -> Happy_IntList
                                                               -> HappyStk HappyAbsSyn
                                                                  -> P HappyAbsSyn)
```

* Finally, commit c00e6dc96ba9ad694ea8298e6a8b2fb99c75634c introduces the current (buggy) behaviour:

```
$ happy -a -o compileBug.hs compileBug.y
$ ghc compileBug.hs
[1 of 1] Compiling Main             ( compileBug.hs, compileBug.o )

compileBug.hs:61:17: error:
    • Couldn't match type ‘t4’ with ‘()’
      ‘t4’ is a rigid type variable bound by
        the type signature for:
          happyReduce_1 :: forall t4.
                           Int
                           -> Token
                           -> Int
                           -> Happy_IntList
                           -> HappyStk (HappyAbsSyn t4)
                           -> P (HappyAbsSyn t4)
        at compileBug.hs:60:18
      Expected type: Int
                     -> Token
                     -> Int
                     -> Happy_IntList
                     -> HappyStk (HappyAbsSyn t4)
                     -> P (HappyAbsSyn t4)
        Actual type: Int
                     -> Token
                     -> Int
                     -> Happy_IntList
                     -> HappyStk (HappyAbsSyn ())
                     -> P (HappyAbsSyn ())
    • In the expression: happySpecReduce_1 0 happyReduction_1
      In an equation for ‘happyReduce_1’:
          happyReduce_1 = happySpecReduce_1 0 happyReduction_1
    • Relevant bindings include
        happyReduce_1 :: Int
                         -> Token
                         -> Int
                         -> Happy_IntList
                         -> HappyStk (HappyAbsSyn t4)
                         -> P (HappyAbsSyn t4)
          (bound at compileBug.hs:61:1)
```

[Here is what happened over the course of those three commits.](https://github.com/simonmar/happy/compare/6ad73f262d2034028c6fa148cee914263c884ccc...c00e6dc96ba9ad694ea8298e6a8b2fb99c75634c)

--- a/happy.cabal
+++ b/happy.cabal
@@ -120,6 +120,7 @@ extra-source-files:
 	tests/precedence002.y
 	tests/test_rules.y
         tests/issue91.y
+        tests/issue94.y
         tests/issue95.y
         tests/monaderror-explist.y
         tests/typeclass_monad001.y

--- a/src/Grammar.lhs
+++ b/src/Grammar.lhs
@@ -32,12 +32,6 @@ Here is our mid-section datatype
 
 > import Control.Monad.Writer
 
-#ifdef DEBUG
-
-> import System.IOExts
-
-#endif
-
 > type Name = Int
 
 > type Production = (Name,[Name],(String,[Int]),Priority)

--- a/src/LALR.lhs
+++ b/src/LALR.lhs
@@ -35,9 +35,24 @@ Generation of LALR parsing tables.
 This means rule $a$, with dot at $b$ (all starting at 0)
 
 > data Lr0Item = Lr0 {-#UNPACK#-}!Int {-#UNPACK#-}!Int                  -- (rule, dot)
->       deriving (Eq,Ord)
+>       deriving (Eq,Ord
+
+#ifdef DEBUG
+
+>       ,Show
+
+#endif
+
+>       )
 
 > data Lr1Item = Lr1 {-#UNPACK#-}!Int {-#UNPACK#-}!Int NameSet  -- (rule, dot, lookahead)
+
+#ifdef DEBUG
+
+>       deriving (Show)
+
+#endif
+
 > type RuleList = [Lr0Item]
 
 -----------------------------------------------------------------------------

--- a/src/ProduceCode.lhs
+++ b/src/ProduceCode.lhs
@@ -107,14 +107,14 @@ Produce the complete output file.
 >                             . str "#endif" . nl
 >                       | otherwise = id
 >
->    n_types = length (filter isNothing (elems nt_types))
+>    n_missing_types = length (filter isNothing (elems nt_types))
 >    happyAbsSyn = str "(HappyAbsSyn " . str wild_tyvars . str ")"
->      where wild_tyvars = unwords (replicate n_types "_")
+>      where wild_tyvars = unwords (replicate n_missing_types "_")
 >
 >    -- This decides how to include (if at all) a type signature
 >    -- See <https://github.com/simonmar/happy/issues/94>
 >    filterTypeSig :: (String -> String) -> String -> String
->    filterTypeSig content | n_types == 0 = content
+>    filterTypeSig content | n_missing_types == 0 = content
 >                          | otherwise = ifGeGhc710 content
 >
 >    top_opts =

--- a/src/ProduceCode.lhs
+++ b/src/ProduceCode.lhs
@@ -14,15 +14,15 @@ The code generator.
 >                                 interleave, interleave', maybestr,
 >                                 brack, brack' )
 
-> import Data.Maybe                     ( isJust, isNothing )
+> import Data.Maybe             ( isJust, isNothing, fromMaybe )
 > import Data.Char
 > import Data.List
 
-> import Control.Monad      ( forM_ )
+> import Control.Monad          ( forM_ )
 > import Control.Monad.ST
-> import Data.Bits          ( setBit )
-> import Data.Array.ST      ( STUArray )
-> import Data.Array.Unboxed ( UArray )
+> import Data.Bits              ( setBit )
+> import Data.Array.ST          ( STUArray )
+> import Data.Array.Unboxed     ( UArray )
 > import Data.Array.MArray
 > import Data.Array.IArray
 
@@ -133,14 +133,14 @@ If we're using coercions, we need to generate the injections etc.
 >             bhappy_item = brack' happy_item
 >
 >             inject n ty
->               = mkHappyIn n . str " :: " . type_param n ty
+>               = mkHappyIn n . str " :: " . typeParam n ty
 >               . str " -> " . bhappy_item . char '\n'
 >               . mkHappyIn n . str " x = Happy_GHC_Exts.unsafeCoerce# x\n"
 >               . str "{-# INLINE " . mkHappyIn n . str " #-}"
 >
 >             extract n ty
 >               = mkHappyOut n . str " :: " . bhappy_item
->               . str " -> " . type_param n ty . char '\n'
+>               . str " -> " . typeParam n ty . char '\n'
 >               . mkHappyOut n . str " x = Happy_GHC_Exts.unsafeCoerce# x\n"
 >               . str "{-# INLINE " . mkHappyOut n . str " #-}"
 >         in
@@ -186,7 +186,7 @@ example where this matters.
 >       . str "\n\t= HappyTerminal " . token
 >       . str "\n\t| HappyErrorToken Int\n"
 >       . interleave "\n"
->         [ str "\t| " . makeAbsSynCon n . strspace . type_param n ty
+>         [ str "\t| " . makeAbsSynCon n . strspace . typeParam n ty
 >         | (n, ty) <- assocs nt_types,
 >           (nt_types_index ! n) == n]
 
@@ -315,7 +315,7 @@ happyMonadReduce to get polymorphic recursion.  Sigh.
 >                 id
 >          else
 >                 nl . reductionFun . strspace
->               . interleave " " (map str (take (length toks) (repeat "_")))
+>               . interleave " " (replicate (length toks) (str "_"))
 >               . str " = notHappyAtAll ")
 
 >     | otherwise
@@ -387,7 +387,7 @@ happyMonadReduce to get polymorphic recursion.  Sigh.
 >               tokLets code''
 >                  | coerce && not (null cases)
 >                       = interleave "\n\t" cases
->                       . code'' . str (take (length cases) (repeat '}'))
+>                       . code'' . str (replicate (length cases) '}')
 >                  | otherwise = code''
 >
 >               cases = [ str "case " . extract t . strspace . mkDummyVar n
@@ -667,13 +667,13 @@ action array indexed by (terminal * last_state) + state
 >       | otherwise
 >           = str "happyActOffsets :: Happy_Data_Array.Array Int Int\n"
 >           . str "happyActOffsets = Happy_Data_Array.listArray (0,"
->               . shows (n_states) . str ") (["
+>               . shows n_states . str ") (["
 >           . interleave' "," (map shows act_offs)
 >           . str "\n\t])\n\n"
 >
 >           . str "happyGotoOffsets :: Happy_Data_Array.Array Int Int\n"
 >           . str "happyGotoOffsets = Happy_Data_Array.listArray (0,"
->               . shows (n_states) . str ") (["
+>               . shows n_states . str ") (["
 >           . interleave' "," (map shows goto_offs)
 >           . str "\n\t])\n\n"
 >           
@@ -683,7 +683,7 @@ action array indexed by (terminal * last_state) + state
 >
 >           . str "happyDefActions :: Happy_Data_Array.Array Int Int\n"
 >           . str "happyDefActions = Happy_Data_Array.listArray (0,"
->               . shows (n_states) . str ") (["
+>               . shows n_states . str ") (["
 >           . interleave' "," (map shows defaults)
 >           . str "\n\t])\n\n"
 >
@@ -764,9 +764,7 @@ outlaw them inside { }
 >                       [ (a, fn a b) | (a, b) <- assocs nt_types ]
 >     where
 >       fn n Nothing = n
->       fn _ (Just a) = case lookup a assoc_list of
->                         Just v -> v
->                         Nothing -> error ("cant find an item in list")
+>       fn _ (Just a) = fromMaybe (error "can't find an item in list") (lookup a assoc_list)
 >       assoc_list = [ (b,a) | (a, Just b) <- assocs nt_types ]
 
 >    makeAbsSynCon = mkAbsSynCon nt_types_index
@@ -1045,8 +1043,8 @@ See notes under "Action Tables" above for some subtleties in this function.
 >                     (act:_) -> act    -- pick the first one we see for now
 >
 >   where reduces
->           =  [ act | (_,act@(LR'Reduce _ _)) <- actions ]
->           ++ [ act | (_,(LR'Multiple _ act@(LR'Reduce _ _))) <- actions ]
+>           =  [ act | (_, act@(LR'Reduce _ _)) <- actions ]
+>           ++ [ act | (_, LR'Multiple _ act@(LR'Reduce _ _)) <- actions ]
 
 -----------------------------------------------------------------------------
 -- Generate packed parsing tables.
@@ -1141,13 +1139,13 @@ See notes under "Action Tables" above for some subtleties in this function.
 >                | (state, acts) <- assocs action,
 >                  let (err:_dummy:vec) = assocs acts
 >                      vec' = drop (n_starts+n_nonterminals) vec
->                      acts' = filter (notFail) (err:vec')
+>                      acts' = filter notFail (err:vec')
 >                      default_act = getDefault acts'
 >                      acts'' = mkActVals acts' default_act
 >                ]
 >
 >        explist_actions :: [(Int, [Int])]
->        explist_actions = [ (state, concat $ map f $ assocs acts)
+>        explist_actions = [ (state, concatMap f $ assocs acts)
 >                          | (state, acts) <- assocs action ]
 >                          where
 >                            f (t, LR'Shift _ _ ) = [t - fst token_names_bound]
@@ -1180,7 +1178,7 @@ See notes under "Action Tables" above for some subtleties in this function.
 >        mkGotoVals assocs' =
 >                [ (token - first_nonterm', i) | (token, Goto i) <- assocs' ]
 >
->        sorted_actions = reverse (sortBy cmp_state (actions++gotos))
+>        sorted_actions = sortBy (flip cmp_state) (actions ++ gotos)
 >        cmp_state (_,_,_,width1,tally1,_) (_,_,_,width2,tally2,_)
 >                | width1 < width2  = LT
 >                | width1 == width2 = compare tally1 tally2
@@ -1377,9 +1375,9 @@ slot is free or not.
 > mkHappyIn n           = str "happyIn"  . shows n
 > mkHappyOut n          = str "happyOut" . shows n
 
-> type_param :: Int -> Maybe String -> ShowS
-> type_param n Nothing   = char 't' . shows n
-> type_param _ (Just ty) = brack ty
+> typeParam :: Int -> Maybe String -> ShowS
+> typeParam n Nothing   = char 't' . shows n
+> typeParam _ (Just ty) = brack ty
 
 > specReduceFun :: Int -> Bool
 > specReduceFun = (<= 3)
@@ -1389,7 +1387,7 @@ slot is free or not.
 -- for placing in a string.
 
 > hexChars :: [Int] -> String
-> hexChars acts = concat (map hexChar acts)
+> hexChars = concatMap hexChar
 
 > hexChar :: Int -> String
 > hexChar i | i < 0 = hexChar (i + 65536)
@@ -1407,7 +1405,7 @@ This guards against integers that are so large as to (when converted using
 larger than an expected minimum value.
 
 > checkedHexChars :: Int -> [Int] -> String
-> checkedHexChars minValue acts = concat (map hexChar' acts)
+> checkedHexChars minValue = concatMap hexChar'
 >   where hexChar' i | checkHexChar minValue i = hexChar i
 >                    | otherwise = error "grammar does not fit in 16-bit representation that is used with '--ghc'"
 

--- a/src/ProduceCode.lhs
+++ b/src/ProduceCode.lhs
@@ -90,16 +90,42 @@ Produce the complete output file.
 >       -- fix, others not so easy, and others would require GHC version
 >       -- #ifdefs.  For now I'm just disabling all of them.
 >
+>    partTySigs_opts = ifGeGhc710 (str "{-# OPTIONS_GHC -XPartialTypeSignatures #-}" . nl)
+>
 >    intMaybeHash | ghc       = str "Happy_GHC_Exts.Int#"
 >                 | otherwise = str "Int"
 >
->    top_opts = nowarn_opts .
->      case top_options of
->          "" -> str ""
->          _  -> str (unwords [ "{-# OPTIONS"
->                             , top_options
->                             , "#-}"
->                             ]) . nl
+>    -- Parsing monad and its constraints
+>    pty = str monad_tycon
+>    pcont = str monad_context
+>
+>    -- If GHC is enabled, wrap the content in a CPP ifdef that includes the
+>    -- content and tests whether the GHC version is >= 7.10.3
+>    ifGeGhc710 :: (String -> String) -> String -> String
+>    ifGeGhc710 content | ghc = str "#if __GLASGOW_HASKELL__ >= 710" . nl
+>                             . content
+>                             . str "#endif" . nl
+>                       | otherwise = id
+>
+>    n_types = length (filter isNothing (elems nt_types))
+>    happyAbsSyn = str "(HappyAbsSyn " . str wild_tyvars . str ")"
+>      where wild_tyvars = unwords (replicate n_types "_")
+>
+>    -- This decides how to include (if at all) a type signature
+>    -- See <https://github.com/simonmar/happy/issues/94>
+>    filterTypeSig :: (String -> String) -> String -> String
+>    filterTypeSig content | n_types == 0 = content
+>                          | otherwise = ifGeGhc710 content
+>
+>    top_opts =
+>        nowarn_opts
+>      . (case top_options of
+>           "" -> str ""
+>           _  -> str (unwords [ "{-# OPTIONS"
+>                              , top_options
+>                              , "#-}"
+>                              ]) . nl)
+>      . partTySigs_opts
 
 %-----------------------------------------------------------------------------
 Make the abstract syntax type declaration, of the form:
@@ -341,14 +367,7 @@ happyMonadReduce to get polymorphic recursion.  Sigh.
 >                           | otherwise                  = nt
 >
 >               mkReductionHdr lt' s =
->                       let pcont = str monad_context
->                           pty = str monad_tycon
->                           all_tyvars = [ 't':show n | (n, Nothing) <-
->                                             assocs nt_types ]
->                           str_tyvars = str (unwords all_tyvars)
->                           happyAbsSyn = str "(HappyAbsSyn "
->                                         . str_tyvars . str ")"
->                           tysig = case lexer' of
+>                       let tysig = case lexer' of
 >                             Nothing -> id
 >                             _ | target == TargetArrayBased ->
 >                                 mkReduceFun i . str " :: " . pcont
@@ -359,7 +378,7 @@ happyMonadReduce to get polymorphic recursion.  Sigh.
 >                                 . happyAbsSyn . str " -> "
 >                                 . pty . str " " . happyAbsSyn . str "\n"
 >                               | otherwise -> id in
->                       tysig . mkReduceFun i . str " = "
+>                       filterTypeSig tysig . mkReduceFun i . str " = "
 >                       . str s . strspace . lt' . strspace . showInt adjusted_nt
 >                       . strspace . reductionFun . nl
 >                       . reductionFun . strspace
@@ -427,8 +446,6 @@ The token conversion function.
 >       Just (lexer'',eof') ->
 >       case (target, ghc) of
 >          (TargetHaskell, True) ->
->             let pcont = str monad_context
->                 pty = str monad_tycon in
 >                 str "happyNewToken :: " . pcont . str " => "
 >               . str "(Happy_GHC_Exts.Int#\n"
 >               . str "                   -> Happy_GHC_Exts.Int#\n"
@@ -813,8 +830,6 @@ MonadStuff:
 
 
 >    produceMonadStuff =
->            let pcont = str monad_context
->                pty = str monad_tycon  in
 >            str "happyThen :: " . pcont . str " => " . pty
 >          . str " a -> (a -> "  . pty
 >          . str " b) -> " . pty . str " b\n"
@@ -838,9 +853,6 @@ MonadStuff:
 >                . errorHandler . str "\n"
 >               _ ->
 >                let
->                  all_tyvars = [ 't':show n | (n, Nothing) <- assocs nt_types ]
->                  str_tyvars = str (unwords all_tyvars)
->                  happyAbsSyn = str "(HappyAbsSyn " . str_tyvars . str ")"
 >                  happyParseSig
 >                    | target == TargetArrayBased =
 >                      str "happyParse :: " . pcont . str " => " . intMaybeHash
@@ -871,7 +883,7 @@ MonadStuff:
 >                      . str " -> " . pty . str " " . happyAbsSyn . str ")\n"
 >                      . str "\n"
 >                    | otherwise = id in
->                  happyParseSig . newTokenSig . doActionSig . reduceArrSig
+>                  filterTypeSig (happyParseSig . newTokenSig . doActionSig . reduceArrSig)
 >                . str "happyThen1 :: " . pcont . str " => " . pty
 >                . str " a -> (a -> "  . pty
 >                . str " b) -> " . pty . str " b\n"

--- a/templates/GenericTemplate.hs
+++ b/templates/GenericTemplate.hs
@@ -2,9 +2,6 @@
 
 #ifdef HAPPY_GHC
 #undef __GLASGOW_HASKELL__
-#define HAPPY_IF_GHC_GT_500 #if __GLASGOW_HASKELL__ > 500
-#define HAPPY_IF_GHC_GE_503 #if __GLASGOW_HASKELL__ >= 503
-#define HAPPY_ELIF_GHC_500 #elif __GLASGOW_HASKELL__ == 500
 #define HAPPY_IF_GHC_GT_706 #if __GLASGOW_HASKELL__ > 706
 #define HAPPY_ELSE #else
 #define HAPPY_ENDIF #endif

--- a/tests/Makefile
+++ b/tests/Makefile
@@ -19,7 +19,7 @@ endif
 
 TESTS = Test.ly TestMulti.ly TestPrecedence.ly bug001.ly \
 	monad001.y monad002.ly precedence001.ly precedence002.y \
-	bogus-token.y bug002.y Partial.ly issue91.y issue95.y \
+	bogus-token.y bug002.y Partial.ly issue91.y issue94.y issue95.y \
 	AttrGrammar001.y AttrGrammar002.y \
 	test_rules.y monaderror.y monaderror-explist.y \
 	typeclass_monad001.y typeclass_monad002.ly typeclass_monad_lexer.y

--- a/tests/issue94.y
+++ b/tests/issue94.y
@@ -1,0 +1,33 @@
+-- See <https://github.com/simonmar/happy/issues/94> for more information
+%name parse prod
+
+%tokentype { Token }
+
+%monad { P } { bindP } { returnP }
+%error { error "parse error" }
+%lexer { lexer } { EOF }
+
+%token
+  IDENT  { Identifier $$ }
+
+%%
+
+prod
+  : IDENT { () }
+
+{
+data Token = EOF | Identifier String
+
+type P a = String -> (a, String)
+
+bindP :: P a -> (a -> P b) -> P b
+bindP p f s = let (x,s') = p s in f x s'
+
+returnP :: a -> P a
+returnP = (,)
+
+lexer :: (Token -> P a) -> P a
+lexer cont s = cont (case s of { "" -> EOF; _ -> Identifier s }) ""
+
+main = return ()
+}

